### PR TITLE
Feature/implementation

### LINKED ISSUE #2: [GPC] Filter
## Summary
The precompiler would only include the elements of a feature file in the result, which match a [cucumber-tag-expression](https://www.npmjs.com/package/cucumber-tag-expressions) (e.g., has a given tag, does not have a given tag) 

## Use cases
The base feature file set contains all scenarios, including WIP or obsolete ones, for different suites and so on. The resulting feature files would only contain elements relevant to the test run to make the feature files more transparent.

## Example
Take the following (example) feature file:
```gherkin
# src/login.feature
Feature: Login

  Scenario: Login pass
    Given the login page is opened
    When the username is typed
    And the password is typed
    And the login button is clicked
    Then the user should be logged in

  @wip
  Scenario: Login fail
    Given the login page is opened
    When the username is typed
    And the wrong password is typed
    And the login button is clicked
    Then the error should be visible
    And the user should not be logged in
```

Running the precompiled would result in the following feature files when executed with `not @wip` expression:
```gherkin
# dist/login.feature
Feature: Login

  Scenario: Login pass
    Given the login page is opened
    When the username is typed
    And the password is typed
    And the login button is clicked
    Then the user should be logged in
```

## Options/Variants

- Keeping feature files if there is no element matching the expression in it.

## Additional context
Related issue: gherking/gherking#39

### PR DESCRIPTION
**Pull request by rakur**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,25 +30,63 @@ await compiler.save('./features/dist/login.feature', ast, {
...
+
+| Option | Type | Description | Necessity |
+|:------:|:----:|:------------|:----------|
+| `includeScenarioTags` | `Array` | Scenarios with the given tag will be included in the result | Optional |
+| `excludeScenarioTags` | `Array` | Scenarios with the given tag will be excluded from the result | Optional |
+| `includeRuleTags` | `Array` | Rules with the given tag will be included in the result | Optional |
+| `excludeRuleTags` | `Array` | Rules with the given tag and all their elements will be excluded from the result | Optional |
+
+## Example
+Excluding scenarios with @wip tag
```
Comment (szikszail) [inline README.md]: Please add the configuration of the example as well, that would demonstrate the usage of cucumber tag expressions.

File: src/types.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+export interface FilterConfig {
+    excludeScenarioTags?: Array<string>,
```
Comment (szikszail) [inline src/types.ts]: Please use `string[]`, in general in the source files.

File: src/types.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface FilterConfig {
```
Comment (szikszail) [inline src/types.ts]: In the description of the GPC, you mention the use of cucumber-tag-expressions, but you seem to expect individual tags and not the expression. I'd suggest updating the implementation to use the expression (https://www.npmjs.com/package/cucumber-tag-expressions).

File: src/index.ts [hunk 4]
```diff
@@ -1,28 +1,120 @@
...
+    }
+    if (config.includeRuleTags && config.excludeRuleTags) {
+        invalidRuleTags = checkDuplication(config.includeRuleTags, config.excludeRuleTags)
+    }
+    if (invalidScenarioTags.length || invalidRuleTags.length) {
+        throw new Error(`The following tag(s) are present as both inclusion and exclusion: ${invalidScenarioTags.concat(invalidRuleTags)}`);
+    }
+}
+
+class Filter implements PreCompiler {
```
Comment (szikszail) [inline src/index.ts:18]: In general, you implemented the filtering logic at the event method level. I think the pre methods would be more suitable for this logic.

**Review by szikszail — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 5]
```diff
@@ -1,28 +1,97 @@
...
+        }
+        return false;
+    }
 
-// TODO: Add implementation of your precompiler
-class Template implements PreCompiler {
-    constructor() {
-        debug("Intialize");
+    postFeature(feature: Feature): boolean {
+        if (feature.elements.length && feature.elements.some((element: Element) =>
```
Comment (szikszail) [inline src/index.ts]: Feature files can be localized, thus relying on the keywords is not suggested. Please use `instanceof`. 

Also:
```javascript
if (condition) {
  return true;
}
return false;
// instead:
return condition;
```

File: src/index.ts [hunk 6]
```diff
@@ -1,28 +1,97 @@
...
+        }
+        if (parent instanceof Rule) {
+            tags = tags.concat(tagsToString(parent.tags));
+        }
+
+        return this.parsedFilter.evaluate(tags);
+    }
+
+    postRule(rule: Rule): boolean {
+        if (rule.elements.length && rule.elements.some((element: Element) => element.keyword === "Scenario" || element.keyword === "Scenario Outline")) {
```
Comment (szikszail) [inline src/index.ts]: Same as for the `postFeature`

File: src/index.ts [hunk 7]
```diff
@@ -1,28 +1,97 @@
...
+                        .concat(tagsToString(this.featureTags));
+            return this.parsedFilter.evaluate(exampleTags);
+        });
+        if (!scenarioOutline.examples.length) {
+            return false;
+        }  
+        let tags = tagsToString(scenarioOutline.tags)
+                        .concat(tagsToString(this.featureTags));
+        
+        for (const example of scenarioOutline.examples) {
```
Comment (szikszail) [inline src/index.ts]: You have already evaluated at this point that there is at least one example matching the expression, thus I don't think you need to check example tags, or even checking tags of the scenario outline seems unnecessary at this point.

File: README.md [hunk 8]
```diff
@@ -30,25 +30,65 @@ await compiler.save('./features/dist/login.feature', ast, {
...
 });
 ```
+
+## Configuration
+
+The precompiler accepts the following configuration:
+
+| Option | Type | Description | Necessity |
+|:------:|:----:|:------------|:----------|
+| `includeScenarioTags` | `Array` | Scenarios with the given tag will be included in the result | Optional |
```
Comment (szikszail) [inline README.md]: Please update this table.

File: README.md [hunk 9]
```diff
@@ -30,25 +30,65 @@ await compiler.save('./features/dist/login.feature', ast, {
...
+| `excludeScenarioTags` | `Array` | Scenarios with the given tag will be excluded from the result | Optional |
+| `includeRuleTags` | `Array` | Rules with the given tag will be included in the result | Optional |
+| `excludeRuleTags` | `Array` | Rules with the given tag and all their elements will be excluded from the result | Optional |
+
+## Example
+Excluding scenarios with @wip or @exclude tags
```
Comment (szikszail) [inline README.md]: I'd either rephrase it: "Keeping scenarios with `@current` tag", or set a filter what means this condition.

**Review by szikszail — CHANGES_REQUESTED** [review]

**Review by szikszail — APPROVED** [review]

**Review by MonikaLaszlo — APPROVED** [review]

**Pull request merged**